### ISSUE TITLE
Update Grub Menus does not correctly display lvm

### ISSUE THREAD

**Issue by cjg67** [rescatux/rescapp#53]
Hello Adrian,

I have just moved to using lvm2 to manage my disk drives, and whilst testing resatux I accidently clicked on the update grub menus and when it completed, I noticed the following under file system:

No-file-system

It correctly identifies ext4 but fails to list lvm as the type.

If you go to blkid the following is listed:

`root@debian:/home/user# blkid                                                                                                                                
/dev/sr0: UUID="2019-12-02-21-12-14-00" LABEL="Rescatux0.72-beta5" TYPE="iso9660" PTUUID="387e3b04" PTTYPE="dos"                                             
/dev/loop0: TYPE="squashfs"                                                                                                                                  
/dev/sda1: UUID="9bd5a9e0-74b9-4b6b-b249-fdafbd8f4d95" TYPE="ext2" PARTUUID="cc8932fa-01"                                                                    
/dev/sda5: UUID="f71ixK-gWeQ-zAhw-VPcf-yV9U-vjxo-DvDHWB" TYPE="LVM2_member" PARTUUID="cc8932fa-05"                                                           
/dev/sdb: UUID="Kd2Ovb-BZZd-7d1z-al8J-4Zji-sQJD-X11On1" TYPE="LVM2_member"                                                                                   
/dev/sdc: UUID="vjt3dl-XdaU-qtMM-zVdx-CIvK-mDpC-IXcjXG" TYPE="LVM2_member"                                                                                   
/dev/sdd1: PARTUUID="0e6cd98f-0f06-5549-b4eb-6180ee01dc07"                                                                                                   
/dev/sdd2: UUID="22c49e48-2918-458b-a2d4-0e1056dd08c3" TYPE="ext4" PARTUUID="fb45a6ae-6863-104e-bf1d-6f6b8ef1f6d8"                                           
/dev/sdd3: UUID="ab48ffb0-d551-4476-aae2-b70b80a12610" TYPE="swap" PARTUUID="3e0e1c1b-495c-f34e-b3be-0e800c036e3a"                                           
/dev/sdd4: UUID="d6a57af3-dc62-43af-ae69-331c57a9b1cf" TYPE="ext4" PARTUUID="8b774b3a-abc2-6b4d-b230-e294a6da64dc"                                           
/dev/sde: UUID="SVEcvG-ufOM-20q4-8vMP-d8E2-dWDU-HYAkqs" TYPE="LVM2_member"                                                                                   
/dev/sdf: PTUUID="2f02146f-aaa6-4db6-9686-77123d32af45" PTTYPE="gpt"                                                                                         
/dev/mapper/debian--vg-root: UUID="94ca3e5f-58a4-42b4-9795-0ae31ab4eb73" TYPE="ext4"                                                                         
/dev/mapper/debian--vg-swap_1: UUID="826ad692-ad8a-411a-b1a5-16fe03427a66" TYPE="swap"                                                                       
/dev/mapper/lfs--lvm-home: UUID="ed437848-f762-432a-8bbd-0b42a06a45ce" TYPE="ext4"                                                                           
/dev/mapper/lfs--lvm-root: UUID="67419c73-11d1-4132-bc23-a3d62576985b" TYPE="ext4"                                                                           
/dev/mapper/lfs--lvm-swap: UUID="ae5e314e-6c7e-486d-8259-d36740f4842f" TYPE="swap"`

Gparted also correctly shows the type as lvm.  I have no idea as to what is going wrong here, but I really do believe that this is a bug.  If it is a grub tool, then it may well be beyond "fixing" as it is probably meant to do what it is.

On linuxfromscratch we do not use os-prober or a tool to generate a grub.cnf file, as the developers early on found that it is not compatible with the structure of lfs/blfs.

Regards,

Christopher.

**Comment by adrian15** [comment]
I don't think this is a bug but rather an enhacement.
When you say that you noticed _No-file-system_ do you mean you saw it in some of the Gnu/Linux partition screens where you were asked to pick a Gnu/Linux partition?

Can you please describe on text what was on the LVM row?
Or can you please attach an screenshot or photo?

If I'm not mistaken the partition name would have been identified as LV1 or VG1-LV1 where VG1 is its Volume Group and LV1 is its Logical Group.

And depending on which column you are complaining about the current filesystem detection tool/script wouldn't be able to inspect inside the LV to detect its inner ext4.

In summary I would appreciate a text description or a summary so that I can save trying to replicate in my virtual machines for now and giving you a more accurate answer.

Thank you.

**Comment by adrian15** [comment]
> On linuxfromscratch we do not use os-prober or a tool to generate a grub.cnf file, as the developers early on found that it is not compatible with the structure of lfs/blfs.

What generates grub.conf file in most distributions is either grub-mkconfig, grub2-mkconfig or the update-grub wrapper around it. And that's what Rescapp uses under the hood asking it to the distro chroot.

Then, if I'm not mistaken grub-mkconfig can optionally use os-prober (if it detects it to be available) to make nicer menus (detect with more accuracy the present OSes on your system).

My question is: Can I rely in grub-mkconfig being available in a Linux From Scratch typical installation in order to perform the rescapp _Update Grub_ option?

**Comment by cjg67** [comment]
Hello Adrian,

When the update grub menus finishes, on the list of operating systems detected, at the top there is a header called file-system in that column it has for example, Arch linux, and in the file system type it has ext4.  The next operating system detected is Debian and in the file system type that is where it states No-file-system.

I will reboot in a bit back into rescatux and get an exact screen shot.

With regards to grub-mkconfig it is most definately installed on linuxfromscratch.  I just double checked.

Regards,

Christopher.

**Comment by cjg67** [comment]
Hello Adrian,

Here is the screenshot:

![test](https://user-images.githubusercontent.com/50721017/70128091-320dbd00-1674-11ea-8590-013c938dc225.jpg)

Regards,

Christopher.

**Comment by adrian15** [comment]
So here there is the helper that handles the filesystem labeling: https://github.com/rescatux/rescapp/blob/f4e61121a471a2391730c05e05275d624e7a76cb/helpers/rescapp-show-partition-filesystem .

You can run it from the CLI on Rescatux disk by doing:
```
rescapp-show-partition-filesystem /dev/sda1
```
.

As you can see in its source sda1 is converted into sda so that the parted libraries can then check the first partition on the sda disk.

Your LVM named "dm-0" is stripped into "dm-" and then "dm-" does not happen to be a hard disk and thus a filesystem cannot be found.

This code should be reworked quite a bit to support LVs properly.
[pyparted](https://github.com/dcantrell/pyparted) does not seem to have a complete LV support so that might be more difficult than initially thought.

If you find another method/script that given an LV path or LV name then gives you its associate filesystem please share it here and I'm sure I will be able to make something out of it.

Thank you.

**Comment by cjg67** [comment]
Hello Adrian,

I have managed to find something that *may* work:

lsblk -o TYPE,NAME,KNAME,UUID,MOUNTPOINT,SIZE,FSTYPE

This gives the following output when run:

`TYPE NAME  KNAME UUID                                   MOUNTPOINT   SIZE FSTYPE
disk sda   sda   Kd2Ovb-BZZd-7d1z-al8J-4Zji-sQJD-X11On1            931.5G LVM2_m
lvm  ├─lfs--lvm-home
     │     dm-2  ed437848-f762-432a-8bbd-0b42a06a45ce   /home        500G ext4
lvm  ├─lfs--lvm-root
     │     dm-3  67419c73-11d1-4132-bc23-a3d62576985b   /            200G ext4
lvm  └─lfs--lvm-swap
           dm-4  ae5e314e-6c7e-486d-8259-d36740f4842f   [SWAP]         6G swap
disk sdb   sdb   vjt3dl-XdaU-qtMM-zVdx-CIvK-mDpC-IXcjXG            111.8G LVM2_m
disk sdc   sdc                                                     894.3G 
part ├─sdc1
     │     sdc1                                                        2M 
part ├─sdc2
     │     sdc2  22c49e48-2918-458b-a2d4-0e1056dd08c3              912.7M ext4
part ├─sdc3
     │     sdc3  ab48ffb0-d551-4476-aae2-b70b80a12610                  8G swap
part └─sdc4
           sdc4  d6a57af3-dc62-43af-ae69-331c57a9b1cf              885.3G ext4
disk sdd   sdd   SVEcvG-ufOM-20q4-8vMP-d8E2-dWDU-HYAkqs            931.5G LVM2_m
disk sde   sde                                                     223.6G 
part ├─sde1
     │     sde1  9bd5a9e0-74b9-4b6b-b249-fdafbd8f4d95                243M ext2
part ├─sde2
     │     sde2                                                        1K 
part └─sde5
           sde5  f71ixK-gWeQ-zAhw-VPcf-yV9U-vjxo-DvDHWB            223.3G LVM2_m
lvm    ├─debian--vg-root
       │   dm-0  94ca3e5f-58a4-42b4-9795-0ae31ab4eb73              215.9G ext4
lvm    └─debian--vg-swap_1
           dm-1  826ad692-ad8a-411a-b1a5-16fe03427a66                7.5G swap
disk sdf   sdf                                                     465.8G 
rom  sr0   sr0   2019-12-02-21-12-14-00                              4.4G iso966
disk pktcdvd0
           pktcdvd0
                                                                     4.4G` 

Regards,

Christopher.

**Comment by cjg67** [comment]
Hello Adrian,

This is as a screenshot: 

![Screenshot from 2019-12-05 20-20-09](https://user-images.githubusercontent.com/50721017/70212780-ce32e380-179c-11ea-94c8-4893988daa9f.png)

**Comment by adrian15** [comment]
Ok, this is a draft:

```
#!/bin/bash
npartition="dm-2"
nfilesystem=$(lsblk -o KNAME,FSTYPE | awk '$1 == "'"${npartition}"'" {print $2}')
echo $nfilesystem
```
.

You can save it as a .sh script, give it execution permission and try it so that you can give me some feedback. ;)

**Comment by cjg67** [comment]
Hello Adrian,

That works well.  When run it shows the file system type as ext4.  I found that solution  on serverfault.

It seems that with that command you will be able to display whatever values that you want in the grub menus list script, as it will not do anything silly to the drives.  It just retrieves the information.  

Regards,

Christopher.

**Comment by cjg67** [comment]
Hello Adrian,

Should it not also display the value of the partition as well?  ie dm-2 as well as the fstype of ext4?

Regards,

Christopher.

**Comment by cjg67** [comment]
Hello Adrian,

What I had envisioned when I thought that this was a bug was, that the TYPE, ie lvm, the KNAME, ie dm-2 and the FSTYPE, ie LVM2_member would be displayed when you went to select a drive/partition to update the grub menu on.

Regards,

Christopher.

**Comment by cjg67** [comment]
Hello Adrian,

Another two options that can be used with that command are:

`MAJ:MIN  major:minor device number
PARTFLAGS  partition flags`

Regards,

Christopher.
